Add Territory Sub management tab in Subscriptions

brymut opened the pull request:
## Description

Add Territory sub management in the 'subscriptions' tab in the settings page with a dropdown to switch between stackers & territories. With dropdown menu to unsubscribe from & mute territory.

Closes https://github.com/stackernews/stacker.news/issues/1409

## Screenshots

### Desktop

https://github.com/user-attachments/assets/81f4be92-abf6-4b9f-ba77-4e0190682b31

### Mobile

https://github.com/user-attachments/assets/159cba7c-59c7-4046-bd15-1998defe6ff0



## Additional Context

_Was anything unclear during your work on this PR? Anything we should definitely take a closer look at?_

## Checklist

**Are your changes backwards compatible? Please answer below:**
Y

**On a scale of 1-10 how well and how have you QA'd this change and any features it might affect? Please answer below:**
10

**For frontend changes: Tested on mobile, light and dark mode? Please answer below:**
Y

**Did you introduce any new environment variables? If so, call them out explicitly here:**
No

Soxasora commented:
Hi brymut, I see you're working on this enhancement/feature.
I wanted to give you a tip to avoid linting problems, we have a Development Tips section that also contains [tips for linting](https://github.com/stackernews/stacker.news?tab=readme-ov-file#linting), such as using the [StandardJS VSCode Extension](https://marketplace.visualstudio.com/items?itemName=standard.vscode-standard) to lint as you code. Take a look! ^^

huumn commented on pages/settings/subscriptions/index.js:
```diff
@@ -1,30 +1,72 @@
 import { SubscribeUserContextProvider } from '@/components/subscribeUser'
 
-export const getServerSideProps = getGetServerSideProps({ query: MY_SUBSCRIBED_USERS, authRequired: true })
+export const getServerSideProps = async (ctx) => {
+  const [userRes, subsRes] = await Promise.all([
+    getGetServerSideProps({ query: MY_SUBSCRIBED_USERS, authRequired: true })(ctx),
+    getGetServerSideProps({ query: MY_SUBSCRIBED_SUBS, authRequired: true })(ctx)
+  ])
+  return {
+    props: {
+      ssrUsers: userRes.props.ssrData,
+      ssrSubs: subsRes.props.ssrData
+    }
+  }
+}
```
I'm surprised this worked.

The move here is to make a compound graphql query like `USER_WITH_ITEMS` or make this into different `.js` files to call each query only when it is needed like we do for `/top` path.

huumn commented on pages/settings/subscriptions/index.js:
```diff
@@ -1,30 +1,72 @@
-export default function MySubscribedUsers ({ ssrData }) {
+export default function MySubscribedUsers ({ ssrUsers, ssrSubs }) {
+  const [view, setView] = useState('stackers')
   const subscribeUserContextValue = useMemo(() => ({ refetchQueries: ['MySubscribedUsers'] }), [])
```
This `refetchQueries` pattern should be used for territory subscriptions too.

huumn reviewed (changes requested):
two changes needed

brymut commented:
> two changes needed

Addressed

huumn commented on components/territory-list.js:
```diff
@@ -77,6 +78,10 @@ export default function TerritoryList ({ ssrData, query, variables, destructureD
                     {sub.name}
                   </Link>
                   <Info className='d-flex'><TerritoryInfo sub={sub} /></Info>
+                  <ActionDropdown>
+                    <ToggleSubSubscriptionDropdownItem sub={sub} />
+                    <MuteSubDropdownItem sub={sub} />
+                  </ActionDropdown>
```
There are several pages with territory lists that this does not work on.

huumn reviewed (changes requested):
I fixed an number of things in this PR, some of them subjective, some of them bugs (like extra separators), then I encountered this dropdown bug.

The action dropdown works on the new territory page, but:

1. it's still there when logged out (logged out user can't subscribe/mute)
2. on the `/top/territories/forever` it thinks I'm not subscribed to any territories
3. same as (2) on `/<user nym>/territories`

brymut commented:
> * on the `/top/territories/forever` it thinks I'm not subscribed to any territories

I think for this context and how the `TOP_SUBS` query is being used, we aren't using an authenticated request to the topSubs resolver, so checking for whether or not the user is authenticated wouldn't work here (at least not trivially), so I think the best approach to this issue is to mirror what the top stackers user list is doing by removing the dropdown in this context (not passing the `nymActionDropdown` prop). What do you think?

brymut commented:
> I think that's fine - withholding this new dropdown in contexts where we haven't yet made it work.

ok, I've added a `subActionDropdown` prop for contexts where the user is logged in and as for the subscribing & muting functionality in `userSubs` resolver I've added a way to check the current logged in user's relationship with each sub. This should handle both 1. & 3.

huumn commented on api/resolvers/sub.js:
```diff
@@ -166,10 +166,59 @@ export default {
           OFFSET $4
           LIMIT $5`, ...range, user.id, decodedCursor.offset, limit)
 
+      if (me) {
+        const userSubNames = await models.subSubscription.findMany({
+          where: { userId: me.id },
+          select: { subName: true }
+        })
+        const subscribedSet = new Set(userSubNames.map(s => s.subName))
+        const mutedSubNames = await models.muteSub.findMany({
+          where: { userId: me.id },
+          select: { subName: true }
+        })
+        const mutedSet = new Set(mutedSubNames.map(s => s.subName))
+        subs.forEach(sub => {
+          sub.meSubscription = subscribedSet.has(sub.name)
+          sub.meMuteSub = mutedSet.has(sub.name)
+        })
+      } else {
+        subs.forEach(sub => {
+          sub.meSubscription = false
+          sub.meMuteSub = false
+        })
+      }
+
```
this is not the way we do this normally. when someone is logged in, this resolver now requires 3 roundtrips to the database. normally, we add `meWhatever` fetching to the main query.

huumn commented on pages/[name]/territories.js:
```diff
@@ -25,6 +27,7 @@ export default function UserTerritories ({ ssrData }) {
           query={USER_WITH_SUBS}
           variables={variables}
           destructureData={data => data.userSubs}
+          subActionDropdown={!!me}
```
i wish there were a better way to do this. it's hard to track booleans dependency injected getting passed several components deep.

anyway, just wanted to note this for myself.

brymut replied:
I think I get what you're getting at. Moved the `useMe` hook directly into `TerritoryList`, seems to work fine that way without needing to track the dependency from here.

huumn reviewed (approved):
this all works well now. i think the performance/maintainability of the parts i commented on are a little sub-standard.

i'll probably delay merging until i can fix those if you don't find the time to fix them yourself.
